lib/floe/workflow/states/map.rb

```diff
@@ -77,27 +84,47 @@ def end?
         end
 
         def ready?(context)
-          !context.state_started? || context.state["ItemProcessorContext"].any? { |ctx| item_processor.step_nonblock_ready?(Context.new(ctx)) }
+          !context.state_started? || each_item_processor(context).any? { |ctx| item_processor.step_nonblock_ready?(ctx) }
         end
 
         def wait_until(context)
-          context.state["ItemProcessorContext"].filter_map { |ctx| item_processor.wait_until(Context.new(ctx)) }.min
+          each_item_processor(context).filter_map { |ctx| item_processor.wait_until(ctx) }.min
         end
 
         def waiting?(context)
-          context.state["ItemProcessorContext"].any? { |ctx| item_processor.waiting?(Context.new(ctx)) }
+          each_item_processor(context).any? { |ctx| item_processor.waiting?(ctx) }
         end
 
         def running?(context)
           !ended?(context)
         end
 
         def ended?(context)
-          context.state["ItemProcessorContext"].all? { |ctx| Context.new(ctx).ended? }
+          each_item_processor(context).all?(&:ended?)
+        end
+
+        def failed?(context)
+          contexts = each_item_processor(context)
+
+          # Handle the simple cases first
+          return true  if contexts.all?(&:failed?)
+          return false if contexts.none?(&:failed?)
```

Comment:
Probably for another PR, but feels like these 2 can be dropped.

Not sure what happens if 1 fails and `tolerated_failure_cont` / tolerated_failure_percentage` are not populated. Can we default `tolerated_failure_percentage = 0`?

If that is the case, then we can drop above 2 and all should work?

Reply:
probably punt here

Reply:
I moved some things around so this is a little simpler now